js-practive/js/main.js

```diff
@@ -1 +1,260 @@
+
+    createButton.addEventListener('click', goToUserForm);
+    usersTable.addEventListener('click', onUsersTableClick);
+}
+
+function setHtmlElementUserForm() {
+    userForm = document.getElementById('users-edit');
+    userForm.addEventListener('click', onUserFormClick);
+}
+
+readingHashState();
+
+function readingHashState() {
+    addEventListener('popstate', (event) => {
+            
+        switch (location.hash.split("=")[0]) {
+            case '#user-form': {
+                if (window.location.hash.split("=")[1]) {
+                    fetch('/user-form.html')
+                    .then((res) => res.text())
+                    .then((stringHtml) => document.body.innerHTML = stringHtml)
+                    .then(() => getSingleUser(window.location.hash.split("=")[1]))
+                    .then((user) => {
+                        setHtmlElementUserForm()
+                        setUserData(user)
+                    })
+    
+                } else {
+                    fetch('/user-form.html')
+                    .then((res) => res.text())
+                    .then((stringHtml) => document.body.innerHTML = stringHtml)
+                    .then(() => setHtmlElementUserForm())
+                    .then(() => {
+                        if(getDataOnLocalStorage()) {
+                            setUserData(getDataOnLocalStorage())
+                        }
+                    })
+                }            
+                break
+            }
+            case '': {
+                fetch('')
+                .then((res) => res.text())
+                .then((stringHtml) => document.body.innerHTML = stringHtml)
+                .then(() => {
+                    setUsersTable()
+                    getUsers()
+                });
+            }
+        }
+     });
+}
+
+function getUsers(){
+    fetch('/user')
+    .then((response) =>  response.json())
+    .then(showUsersList)
+}
+
+getUsers()
+
+function showUsersList(data) {
+    const html = data.map(generateUserHTML).join('');
+    usersTable.innerHTML = html;
+}
+
+function generateUserHTML(user) {     
+    return `
+    <tr 
+        class="user"
+        data-id="${user.id}"
+    >
+        <th>${user.fullName}</th>
+        <th>${user.profession}</th>
+        <th>${user.shortInfo}</th>
+        <th>
+            <button class="btn editBtn">Edit</button>
+            <button class="btn deleteBtn">Delete</button>
+        </th>
+    </tr>
+    `
+}
+
+function goToUserForm(id) {
+    if(typeof(id) === 'string') {
+        location.hash = 'user-form=' + id;
+    } else {
+        location.hash = 'user-form';
+    }
+}
+
+function onUserFormClick(e){
+    if (e.target.classList.contains('btn-cancel')){
+        const user = getUserFromForm()
+        setDataOnLocalStorage(user)
+        goToTable()
+    } else if (e.target.classList.contains('btn-save')){
+        saveUserInfo()
+    }
+}
+
+function goToTable() {
```

Comment:
By calling this function i won't always go back. 
BTW it has not sense to wrap only one action in function. Basically it creates more confusion and complexity rather then optimization (or whatever you was trying to reach)/